use esm for requiring default and deprecated features

efx opened the pull request:
See https://github.com/emberjs/data/pull/6231#issuecomment-510088422 for
the motivation.

~~`yarn test` blows up locally, which might be related to not properly typing the import of `default-features.js` or not correctly exporting it in `@ember/canary-features`.~~

@rwjblue if you have free moments to give me pointers / review here that would be great.

eflanagan0 committed:
attempt to use esm for requiring features

See https://github.com/emberjs/data/pull/6231#issuecomment-510088422 for
the motivation.

efx commented on broccoli/features.js:
```diff
@@ -1,35 +1,12 @@
-    ts.ScriptTarget.ES2017,
-    /*setParentNodes */ true
+  const { default: features } = requireEsm(
+    '../packages/@ember/canary-features/default-features.js'
```
I could not use `@ember/canary-features/default-features.js` for the path. Would this be because ember.js' internal package structure differs from ember data?

rwjblue replied:
I think the local require (e.g. starting with `../packages/`) is totally correct here.

efx commented on packages/@ember/canary-features/index.ts:
```diff
@@ -12,19 +12,8 @@ import { assign } from '@ember/polyfills';
-  EMBER_FRAMEWORK_OBJECT_OWNER_ARGUMENT: true,
-  EMBER_GLIMMER_SET_COMPONENT_TEMPLATE: null,
-};
+import { default: DEFAULT_FEATURES } from '@ember/canary-features/default-features';
```
tsc errors for these lines:
```
$ tsc --noEmit
packages/@ember/canary-features/index.ts:15:10 - error TS1003: Identifier expected.

15 import { default: DEFAULT_FEATURES } from '@ember/canary-features/default-features';
            ~~~~~~~

packages/@ember/canary-features/index.ts:15:17 - error TS1005: ',' expected.

15 import { default: DEFAULT_FEATURES } from '@ember/canary-features/default-features';
```
How can I declare the type while importing? i.e. apart from embedding it into the `default-features.js`

efx replied:
So I think I need to add a [declaration](https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html#objects-with-properties).

rwjblue replied:
Ya, I think you can do:

```js
// packages/@ember/canary-features/default-features.d.ts

export default {
  [key: string]: null | boolean;
}
```

efx commented on broccoli/features.js:
```diff
@@ -1,35 +1,12 @@
 'use strict';
 
-const fs = require('fs');
-const ts = require('./typescript');
```
I see the deprecated features also relies upon this module; I can remove it and merge the deprecated features functionality with this new approach once I sort how the typescript particulars below.

rwjblue replied:
Ya, sounds good to me.

rwjblue commented:
Thanks for working on this @efx!

efx committed:
Update packages/@ember/canary-features/index.ts

Co-Authored-By: Robert Jackson <me@rwjblue.com>

eflanagan0 committed:
redeclare and export to fix typescript linting

eflanagan0 committed:
add a loose declaration of default features hash

efx commented on packages/@ember/canary-features/default-features.d.ts:
```diff
@@ -0,0 +1,2 @@
+declare let DEFAULT_FEATURES: any;
```
A permissive declaration. What would be the appropriate type definition here? A namespace or interface?

rwjblue replied:
Ya, I think you can do:

```js
// packages/@ember/canary-features/default-features.d.ts

export default {
  [key: string]: null | boolean;
}
```

efx replied:
Entering that snippet exactly throws some more typescript errors:
```
 yarn lint:tsc
yarn run v1.17.3
$ tsc --noEmit
packages/@ember/canary-features/default-features.d.ts:2:7 - error TS1005: ']' expected.

2   [key: string]:  null | boolean;
        ~

packages/@ember/canary-features/default-features.d.ts:2:15 - error TS1005: ',' expected.

2   [key: string]:  null | boolean;
                ~

packages/@ember/canary-features/default-features.d.ts:2:16 - error TS1136: Property assignment expected.

2   [key: string]:  null | boolean;
                 ~

packages/@ember/canary-features/default-features.d.ts:2:24 - error TS1005: ':' expected.

2   [key: string]:  null | boolean;
                         ~

packages/@ember/canary-features/default-features.d.ts:2:33 - error TS1005: ',' expected.

2   [key: string]:  null | boolean;
```
I'll keep muddling with the syntax in the meantime. :)

rwjblue replied:
Whoops, have to use `interface` keyword here:

```ts
export default interface DEFAULT_FEATURES {
  [key: string]: null | boolean;
}
```

efx replied:
I think I found the correct combination. I needed to declare an interface and then export a variable that implemented it.

rwjblue commented on packages/@ember/canary-features/index.ts:
```diff
@@ -12,19 +12,8 @@ import { assign } from '@ember/polyfills';
-  EMBER_FRAMEWORK_OBJECT_OWNER_ARGUMENT: true,
-  EMBER_GLIMMER_SET_COMPONENT_TEMPLATE: null,
-};
+import defaultFeatures from '@ember/canary-features/default-features';
```
```suggestion
export { default as DEFAULT_FEATURES } from './default-features';

import DEFAULT_FEATURES from './default-features';
```

rwjblue commented on packages/@ember/canary-features/index.ts:
```diff
@@ -12,19 +12,8 @@ import { assign } from '@ember/polyfills';
-  EMBER_GLIMMER_SET_COMPONENT_TEMPLATE: null,
-};
+import defaultFeatures from '@ember/canary-features/default-features';
+export const DEFAULT_FEATURES = defaultFeatures;
```
```suggestion
```

efx committed:
Update packages/@ember/canary-features/index.ts

Co-Authored-By: Robert Jackson <me@rwjblue.com>

efx committed:
Update packages/@ember/canary-features/index.ts

Co-Authored-By: Robert Jackson <me@rwjblue.com>

eflanagan0 committed:
tighten the type declaration for default features

Made a breakthrough reading https://basarat.gitbooks.io/typescript/content/docs/types/ambient/variables.html

efx commented:
I've updated the type but it still complains about exporting an interface instead of avalue
<img width="987" alt="Screen Shot 2019-07-25 at 13 44 46" src="https://user-images.githubusercontent.com/472721/61896176-6319fc00-aee2-11e9-9421-dc2d3bacaf73.png">

eflanagan0 committed:
use esm for consuming deprecated features

eflanagan0 committed:
merge with master

Add newly deprecated feature from #18214

eflanagan0 committed:
Merge branch 'master' into canary-feature-flags-esm

rwjblue commented on packages/@ember/deprecated-features/deprecated-features.js:
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable no-implicit-coercion */
```
Lets move this to be the `packages/@ember/deprecated-features/index.js` (along with the `packages/@ember/deprecated-features/deprecated-features.d.ts` -> `packages/@ember/deprecated-features/index.d.ts`).

eflanagan0 committed:
rename to index

eflanagan0 committed:
Merge branch 'master' into canary-feature-flags-esm

rwjblue commented:
<details>
<summary>Currently some linting failures</summary>

```
packages/@ember/-internals/environment/lib/env.ts:1:10 - error TS2305: Module '"../../../deprecated-features"' has no exported member 'FUNCTION_PROTOTYPE_EXTENSIONS'.
304
3051 import { FUNCTION_PROTOTYPE_EXTENSIONS } from '@ember/deprecated-features';
306           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
307
308packages/@ember/-internals/glimmer/lib/utils/custom-component-manager.ts:3:10 - error TS2305: Module '"../../../../deprecated-features"' has no exported member 'COMPONENT_MANAGER_STRING_LOOKUP'.
309
3103 import { COMPONENT_MANAGER_STRING_LOOKUP } from '@ember/deprecated-features';
311           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
312
313packages/@ember/-internals/metal/lib/mixin.ts:18:10 - error TS2305: Module '"../../../deprecated-features"' has no exported member 'ALIAS_METHOD'.
314
31518 import { ALIAS_METHOD } from '@ember/deprecated-features';
316            ~~~~~~~~~~~~
317
318packages/@ember/-internals/routing/lib/system/route.ts:23:10 - error TS2305: Module '"../../../../deprecated-features"' has no exported member 'ROUTER_EVENTS'.
319
32023 import { ROUTER_EVENTS } from '@ember/deprecated-features';
321            ~~~~~~~~~~~~~
322
323packages/@ember/-internals/routing/lib/system/router.ts:5:10 - error TS2305: Module '"../../../../deprecated-features"' has no exported member 'APP_CTRL_ROUTER_PROPS'.
324
3255 import { APP_CTRL_ROUTER_PROPS, ROUTER_EVENTS } from '@ember/deprecated-features';
326           ~~~~~~~~~~~~~~~~~~~~~
327
328packages/@ember/-internals/routing/lib/system/router.ts:5:33 - error TS2305: Module '"../../../../deprecated-features"' has no exported member 'ROUTER_EVENTS'.
329
3305 import { APP_CTRL_ROUTER_PROPS, ROUTER_EVENTS } from '@ember/deprecated-features';
331                                  ~~~~~~~~~~~~~
332
333packages/@ember/polyfills/index.ts:1:10 - error TS2305: Module '"../deprecated-features"' has no exported member 'MERGE'.
334
3351 import { MERGE } from '@ember/deprecated-features';
336           ~~~~~
337
338packages/ember-template-compiler/lib/plugins/deprecate-send-action.ts:3:10 - error TS2305: Module '"../../../@ember/deprecated-features"' has no exported member 'SEND_ACTION'.
339
3403 import { SEND_ACTION } from '@ember/deprecated-features';
341           ~~~~~~~~~~~
342
343packages/ember-template-compiler/lib/plugins/index.ts:23:10 - error TS2305: Module '"../../../@ember/deprecated-features"' has no exported member 'SEND_ACTION'.
344
34523 import { SEND_ACTION } from '@ember/deprecated-features';
346            ~~~~~~~~~~~
347
348
349Found 9 errors.
```
</details>

And some legit errors in the test run, e.g.:

```
this._paramsFor is not a function
```

efx commented:
Hmm, I thought my reexport was working but clearly it does not. Not sure about the exact syntax to export those const declarations from the `index.ts` so I can avoid repeating what is in `index.js`. I'll try to finish that later today or this weekend.

eflanagan0 committed:
copy and paste to fix typescript errors

*I know not typescript*
This makes it work but now there are 2 copies of these flags in
`index.js` and `index.ts`. I am pushing up to avoid blocking and we can
consider a follow up PR or any intervening TS guidance to solve the
duplication.

efx commented:
Closing this as I've run out of steam. I might come back to it once I actually learn TS. May need to spike and try `ts-node` as esm doesn't interoperate directly with TS.
